stdlib/public/core/While.swift.gyb

```diff
@@ -0,0 +1,395 @@
+//===--- While.swift.gyb - Lazy views for prefix(while:) & drop(while:) ---===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+%{
+from gyb_stdlib_support import (
+    TRAVERSALS,
+    collectionForTraversal
+)
+}%
+
+//===----------------------------------------------------------------------===//
+// prefix(while:)
+//===----------------------------------------------------------------------===//
+
+/// An iterator over the initial elements traversed by a base iterator that
+/// satisfy a given predicate.
+///
+/// This is the associated iterator for the `LazyPrefixWhileSequence`,
+/// `LazyPrefixWhileCollection`, and `LazyPrefixWhileBidirectionalCollection`
+/// types.
+public struct LazyPrefixWhileIterator<Base : IteratorProtocol> :
+  IteratorProtocol, Sequence {
+
+  public mutating func next() -> Base.Element? {
+    // Return elements from the base iterator until one fails the predicate.
+    if !_predicateHasFailed, let nextElement = _base.next() {
+      if _predicate(nextElement) {
+        return nextElement
+      } else {
+        _predicateHasFailed = true
+      }
+    }
+    return nil
+  }
+
+  internal init(_base: Base, predicate: (Base.Element) -> Bool) {
+    self._base = _base
+    self._predicate = predicate
+  }
+
+  public var base: Base { return _base }
```

Comment:
If you choose to make `base` public, please add tests that check which state `base` is in before and after each `next()` call, because users who use `base` will rely on the specific element `base` will be pointing to.